ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -697,21 +701,117 @@ public void findEntriesSinceOnRestartTest() throws IOException, StoreException {
   }
 
   /**
+   * Tests behaviour of {@link PersistentIndex#findEntriesSince(FindToken, long)} on crash restart for the following
+   * scenario
+   * After restart, lets say no new writes have gone into the store.
+   * For a findEntriesSince(offset beyond logEndOffsetOnStartup with different session id) call, index will reset the
+   * token to logEndOffsetOnStartup and returns the same.
+   * On the subsequent findEntriesSince() call, the index should start returning entries
+   * starting from that offset and should not consider that token as non-inclusive
+   * @throws IOException
+   * @throws StoreException
+   */
+  @Test
+  public void findEntriesSinceOnCrashRestartTest() throws IOException, StoreException {
+    UUID oldSessionId = sessionId;
+    addPutEntries(3, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    reloadIndex(true);
+    // create a token that will be past the index end offset on start up after restart.
+    StoreFindToken startToken = new StoreFindToken(
+        new Offset(log.getEndOffset().getName(), (index.getCurrentEndOffset().getOffset() + (2 * PUT_RECORD_SIZE))),
+        oldSessionId, incarnationId, false);
+    // end token should point to end offset of last entry i.e. (index.getCurrentEndOffset().getOffset() + PUT_RECORD_SIZE)
+    long bytesRead =
+        log.getDifference(new Offset(log.getEndOffset().getName(), index.getCurrentEndOffset().getOffset()),
+            new Offset(log.getFirstSegment().getName(), 0));
+    StoreFindToken expectedEndToken =
+        new StoreFindToken(new Offset(log.getEndOffset().getName(), index.getCurrentEndOffset().getOffset()), sessionId,
+            incarnationId, true);
+    expectedEndToken.setBytesRead(bytesRead);
+    // Fetch the FindToken returned from findEntriesSince
+    FindInfo findInfo = index.findEntriesSince(startToken, Long.MAX_VALUE);
+    assertEquals("EndToken mismatch ", expectedEndToken, findInfo.getFindToken());
+    assertEquals("No entries should have been returned ", 0, findInfo.getMessageEntries().size());
+
+    // add 2 entries to index and log
+    Set<MockId> expectedEntries = new HashSet<>();
+    addPutEntries(1, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    expectedEntries.add(logOrder.lastEntry().getValue().getFirst());
+    addPutEntries(1, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    expectedEntries.add(logOrder.lastEntry().getValue().getFirst());
+    bytesRead = log.getDifference(index.getCurrentEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
+    expectedEndToken = new StoreFindToken(index.journal.getLastOffset(), sessionId, incarnationId, false);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest((StoreFindToken) findInfo.getFindToken(), Long.MAX_VALUE, expectedEntries, expectedEndToken);
+  }
+
+  /**
+   * Tests behaviour of {@link PersistentIndex#findEntriesSince(FindToken, long)} relating to incarnationId
```

Comment:
could you explain how this is testing correctness w.r.t behavior on an incarnation ID change? I might be missing something here but isn't this just a session ID change test? Is the incarnation ID changing? If it isn't then the code that reacts to changes in incarnation ID isn't being exercised, is it? Please clarify.